### PR TITLE
Improve producer ID expiration performance

### PR DESCRIPTION
**Pull request by CalvinLiu7947**
Skip using steam when expiring the producer ID. This can improve the performance significantly when the count is high.
Before
```
Benchmark                                        (numProducerIds)  Mode  Cnt      Score       Error  Units
ProducerStateManagerBench.testDeleteExpiringIds             10000  avgt    3    101.253 ±    28.031  us/op
ProducerStateManagerBench.testDeleteExpiringIds            100000  avgt    3   2297.219 ±  1690.486  us/op
ProducerStateManagerBench.testDeleteExpiringIds           1000000  avgt    3  30688.865 ± 16348.768  us/op
```
After
```
Benchmark                                        (numProducerIds)  Mode  Cnt     Score     Error  Units
ProducerStateManagerBench.testDeleteExpiringIds             10000  avgt    3    39.122 ±   1.151  us/op
ProducerStateManagerBench.testDeleteExpiringIds            100000  avgt    3   464.363 ±  98.857  us/op
ProducerStateManagerBench.testDeleteExpiringIds           1000000  avgt    3  5731.169 ± 674.380  us/op
```
Also, made a change to the JMH testing which excludes the producer ID populating from the testing.

### CODE REVIEW

**Comment by jolshan** [comment]
Thanks Calvin -- similar to this PR https://github.com/apache/kafka/pull/15969
I will take a look soon :)

File: storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java [hunk 1]
```diff
@@ -369,16 +369,21 @@ private boolean isProducerExpired(long currentTimeMs, ProducerStateEntry produce
      * Also expire any verification state entries that are lingering as unverified.
      */
     public void removeExpiredProducers(long currentTimeMs) {
-        List<Long> keys = producers.entrySet().stream()
-                .filter(entry -> isProducerExpired(currentTimeMs, entry.getValue()))
-                .map(Map.Entry::getKey)
-                .collect(Collectors.toList());
+        List<Long> keys = new ArrayList<>();
```
Comment (artemlivshits) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java]: If we decided to do perf improvement, should we also avoid the copy?  I think we can just iterate over the producer id entries and remove expired ones on the spot.

Comment (CalvinLiu7947) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java]: Updated. There is a significant improvement!

**Comment by jolshan** [comment]
Thanks Calvin -- just to be clear, we are also changing the benchmark to not include loading all the producer IDs as well?

**Review by artemlivshits — APPROVED** [review]
LGTM

**Comment by CalvinLiu7947** [comment]
@jolshan Yes, the loading part is excluded. I guess there is not much performance concern for the loading.

**Review by jolshan — APPROVED** [review]
Thanks Calvin!

**Pull request merged**